### PR TITLE
Support HTTP method based allowlists

### PR DESCRIPTION
**Pull request by NickMeves**
Bring over the HTTP Method allowlist & improved validation organization from this defunct PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/685

## Description

Add `--skip-auth-route` option that takes the form `METHOD=PathRegex` or `PathRegex` (for backwards compatible with `--skip-auth-regex` with the eventual intent on deprecating that option).

## Motivation and Context

REST APIs might have very different functionality between a GET vs other HTTP methods that alter state. This supports allowlisting an HTTP method for a given regex path.

E.g. `GET /data/1` can be allowlisted to require no authentication, while `POST /data/1` requires authentication to update the record.

## How Has This Been Tested?

Unit Tests.  TODO - interactive tests

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by NickMeves** [comment]
Leaving this in draft form for now. I mostly shifted parts we wanted to keep from the `RulesEngine` PR.

I want to convert the new `allowlist` tests to Ginkgo before I flip this to ready for review.

File: oauthproxy.go [hunk 1]
```diff
@@ -277,6 +282,51 @@ func buildSignInMessage(opts *options.Options) string {
 	return msg
 }
 
+// buildRoutesAllowlist builds a []route from either the legacy SkipAuthRegex
+// option (paths only support) or newer SkipAuthRoutes option (method=path
+// support)
+func buildRoutesAllowlist(opts *options.Options) ([]*allowedRoute, error) {
```
Comment (NickMeves) [inline oauthproxy.go:288]: I moved configuration out of the options validation (to further the goal of that purely validating and doing no setup).

I know this isn't the best spot, but it gets it with the other very related configuration builders in this file that will eventually be refactored together in another PR to a better spot.

File: pkg/validation/options_test.go [hunk 2]
```diff
@@ -78,12 +77,19 @@ func TestClientSecretFileOption(t *testing.T) {
 	if err != nil {
 		t.Fatalf("failed to create temp file: %v", err)
 	}
-	f.WriteString("testcase")
+	_, err = f.WriteString("testcase")
```
Comment (NickMeves) [inline pkg/validation/options_test.go:80]: I don't know why my IDE catches these when I try to commit but GoSec doesn't 🤷‍♂️ 

These are unrelated to this PR, just handling errors that could be thrown.

File: oauthproxy_test.go [hunk 3]
```diff
@@ -1482,28 +1482,28 @@ func TestAuthOnlyEndpointSetBasicAuthFalseRequestHeaders(t *testing.T) {
 }
 
 func TestAuthSkippedForPreflightRequests(t *testing.T) {
-	upstream := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (NickMeves) [inline oauthproxy_test.go:1485]: naming conflict with imported upstream package -- This is unrelated cleanup that I was warned about when I attempted to commit.

File: oauthproxy.go [hunk 4]
```diff
@@ -277,6 +282,53 @@ func buildSignInMessage(opts *options.Options) string {
...
+		})
+	}
+
+	for _, methodPath := range opts.SkipAuthRoutes {
+		var (
+			method string
+			path   string
+		)
+
+		parts := strings.Split(methodPath, "=")
```
Comment (NickMeves) [inline oauthproxy.go:309]: There's 1 case where this logic won't be fully backwards compatible with migrating former `--skip-auth-regex` entries to this without a method:

Regexes with a `=` in it (e.g querystring) so `/my/path\?q=data` will view `/my/path\?q` as the HTTP method to allow and `data` as the path regex.

The hack for this scenario is `=/my/path\?q=data`.  Just isn't graceful. Forcing a regex wrapper (E.g. `/regex/` or `#regex#` etc) would work? But would make the migration harder for 99% of use cases.

I lean toward leave the corner case not fully compatible, understand its hacky because we are using single configs, and know this will be better with structured configs.

Comment (JoelSpeed) [reply oauthproxy.go:309]: Maybe worth noting this in the changelog?

File: oauthproxy.go [hunk 5]
```diff
@@ -70,6 +76,7 @@ type OAuthProxy struct {
 	AuthOnlyPath      string
 	UserInfoPath      string
 
+	allowedRoutes           []*allowedRoute
```
Comment (JoelSpeed) [inline oauthproxy.go:79]: Does this need to be to be a pointer or could it just be `[]allowedRoute`?

File: oauthproxy.go (hunk 4 above)
Comment (JoelSpeed) [inline oauthproxy.go:309]: What about `strings.SplitN` here? Then you wouldn't need to rejoin the remainder strings later? 

IIRC `SplitN` with `N==2` should produce a maximum of 2 strings, but if the character is not found, will return a slice of length 1 with the original string in it

File: oauthproxy_test.go [hunk 6]
```diff
@@ -2240,6 +2241,237 @@ func TestTrustedIPs(t *testing.T) {
 	}
 }
 
+func Test_buildRoutesAllowlist(t *testing.T) {
+	testCases := []struct {
+		name            string
+		skipAuthRegex   []string
+		skipAuthRoutes  []string
+		expectedMethods []string
+		expectedRegexes []string
```
Comment (JoelSpeed) [inline oauthproxy_test.go:2250]: Could use a dummy struct for this? 

```
type testAllowedRoute struct {
  method string
  regexString string
}

...

assert.Equal(t, route.method, expectedRoute[i].method)
assert.Equal(t, route.regexp.String(), expectedRoute[i].regexString)
```

The benefit of this being that the method and regex are paired, which I would say makes the test easier to read, WDYT?

File: pkg/validation/allowlist.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+	return msgs
+}
+
+// validateRoutes validates method=path routes passed with options.SkipAuthRoutes
+func validateRoutes(o *options.Options) []string {
+	msgs := []string{}
+	for _, route := range o.SkipAuthRoutes {
+		var regex string
+		parts := strings.Split(route, "=")
```
Comment (JoelSpeed) [inline pkg/validation/allowlist.go:35]: Could use `SplitN` here as well

**Review by JoelSpeed — COMMENTED** [review]
Looks great! Left a few suggestions, PTAL

Don't think there's any major changes required here.

Do we have an open issue and/or follow up PR to clean up the `allowed-regex` option?

File: oauthproxy.go (hunk 5 above)
Comment (NickMeves) [reply oauthproxy.go:79]: I weighed the same thing when I was coding this 😄 

Switching it to non-pointer since you brought it up too.  The struct is pretty lightweight already, pointer doesn't buy us much.

**Comment by NickMeves** [comment]
Thanks for the notes @JoelSpeed 

I addressed everything and rebased.

**Review by JoelSpeed — DISMISSED** [review]
Thanks

**Review by JoelSpeed — APPROVED** [review]
